File: tui/src/scala/tui/widgets/ListWidget.scala
package tui
package widgets

import tui.internal.ranges
import tui.internal.saturating._

/** A widget to display several items among which one can be selected (optional)
  *
  * @param block
  * @param items
  * @param style
  *   Style used as a base style for the widget
  * @param startCorner
  * @param highlightStyle
  *   Style used to render selected item
  * @param highlightSymbol
  *   Symbol in front of the selected item (Shift all items to the right)
  * @param repeatHighlightSymbol
  *   Whether to repeat the highlight symbol for each line of the selected item
  */
case class ListWidget(
    state: ListWidget.State = ListWidget.State(),
    items: Array[ListWidget.Item],
    style: Style = Style.DEFAULT,
    startCorner: Corner = Corner.TopLeft,
    highlightStyle: Style = Style.DEFAULT,
    highlightSymbol: Option[String] = None,
    repeatHightlightSymbol: Boolean = false
) extends Widget {

  def getItemsBounds(
      selected0: Option[Int],
      offset0: Int,
      maxHeight: Int
  ): (Int, Int) = {
    val offset = math.min(offset0, items.length.saturating_sub_unsigned(1))
    var start = offset
    var end = offset
    var height = 0
    val it = items.iterator.drop(offset)
    var continue = true
    while (continue && it.hasNext) {
      val item = it.next()
      if (height + item.height > maxHeight) {
        continue = false
      } else {
        height += item.height
        end += 1
      }
    }

    val selected = math.min(selected0.getOrElse(0), items.length - 1)
    while (selected >= end) {
      height = height.saturating_add(items(end).height)
      end += 1
      while (height > maxHeight) {
        height = height.saturating_sub_unsigned(items(start).height)
        start += 1
      }
    }
    while (selected < start) {
      start -= 1
      height = height.saturating_add(items(start).height)
      while (height > maxHeight) {
        end -= 1
        height = height.saturating_sub_unsigned(items(end).height)
      }
    }
    (start, end)
  }

  type State = ListWidget.State

  def render(area: Rect, buf: Buffer): Unit = {
    buf.setStyle(area, style)
    if (area.width < 1 || area.height < 1) {
      return
    }

    if (items.isEmpty) {
      return
    }
    val list_height = area.height

    val (start, end) = getItemsBounds(state.selected, state.offset, list_height)
    state.offset = start

    val highlight_symbol1 = highlightSymbol.getOrElse("")
    val blank_symbol = " ".repeat(Grapheme(highlight_symbol1).width)

    var current_height = 0
    val has_selection = state.selected.isDefined
    ranges.range(state.offset, state.offset + end - start) { i =>
      val item = items(i)
      val (x, y) = startCorner match {
        case Corner.BottomLeft =>
          current_height += item.height
          (area.left, area.bottom - current_height)
        case _ =>
          val pos = (area.left, area.top + current_height)
          current_height += item.height
          pos
      }
      val itemArea = Rect(x, y, width = area.width, height = item.height)

      val item_style = style.patch(item.style)
      buf.setStyle(itemArea, item_style)

      val is_selected = state.selected.contains(i)
      item.content.lines.zipWithIndex.foreach { case (line, j) =>
        // if the item is selected, we need to display the hightlight symbol:
        // - either for the first line of the item only,
        // - or for each line of the item if the appropriate option is set
        val symbol = if (is_selected && (j == 0 || repeatHightlightSymbol)) {
          highlight_symbol1
        } else {
          blank_symbol
        }
        val (elem_x, max_element_width) = if (has_selection) {
          val (elem_x, _) = buf.setStringn(
            x,
            y + j,
            symbol,
            area.width,
            item_style
          )
          (elem_x, area.width - (elem_x - x))
        } else {
          (x, area.width)
        }
        buf.setSpans(elem_x, y + j, line, max_element_width);
      }
      if (is_selected) {
        buf.setStyle(itemArea, highlightStyle)
      }
    }
  }
}

object ListWidget {
  case class State(
      var offset: Int = 0,
      var selected: Option[Int] = None
  ) {
    def select(index: Option[Int]): Unit = {
      selected = index
      if (index.isEmpty) {
        offset = 0
      }
    }
  }

  case class Item(content: Text, style: Style = Style.DEFAULT) {
    def height: Int = content.height
  }
}

File: tui/src/java/tui/widgets/ListWidget.java
package tui.widgets;

import java.util.Optional;
import tui.Buffer;
import tui.Corner;
import tui.Grapheme;
import tui.Position;
import tui.Rect;
import tui.StatefulWidget;
import tui.Style;
import tui.Text;
import tui.Widget;
import tui.internal.Ranges;
import tui.internal.Saturating;

/// A widget to display several items among which one can be selected (optional)
public final class ListWidget implements Widget, StatefulWidget<ListWidget.State> {
  public final Optional<BlockWidget> block;
  public final Item[] items;
  public final Style style;
  public final Corner startCorner;
  public final Style highlightStyle;
  public final Optional<String> highlightSymbol;
  public final boolean repeatHighlightSymbol;

  public ListWidget(
      Optional<BlockWidget> block,
      Item[] items,
      Style style,
      Corner startCorner,
      Style highlightStyle,
      Optional<String> highlightSymbol,
      boolean repeatHighlightSymbol) {
    this.block = block;
    this.items = items;
    this.style = style;
    this.startCorner = startCorner;
    this.highlightStyle = highlightStyle;
    this.highlightSymbol = highlightSymbol;
    this.repeatHighlightSymbol = repeatHighlightSymbol;
  }

  public static ListWidget empty(Item[] items) {
    return new ListWidget(
        Optional.empty(),
        items,
        Style.DEFAULT,
        Corner.TopLeft,
        Style.DEFAULT,
        Optional.empty(),
        false);
  }

  public ListWidget withBlock(BlockWidget b) {
    return new ListWidget(
        Optional.of(b), items, style, startCorner, highlightStyle, highlightSymbol, repeatHighlightSymbol);
  }

  public ListWidget withItems(Item[] i) {
    return new ListWidget(
        block, i, style, startCorner, highlightStyle, highlightSymbol, repeatHighlightSymbol);
  }

  public ListWidget withStyle(Style s) {
    return new ListWidget(
        block, items, s, startCorner, highlightStyle, highlightSymbol, repeatHighlightSymbol);
  }

  public ListWidget withStartCorner(Corner c) {
    return new ListWidget(
        block, items, style, c, highlightStyle, highlightSymbol, repeatHighlightSymbol);
  }

  public ListWidget withHighlightStyle(Style s) {
    return new ListWidget(
        block, items, style, startCorner, s, highlightSymbol, repeatHighlightSymbol);
  }

  public ListWidget withHighlightSymbol(String s) {
    return new ListWidget(
        block, items, style, startCorner, highlightStyle, Optional.of(s), repeatHighlightSymbol);
  }

  public ListWidget withRepeatHighlightSymbol(boolean r) {
    return new ListWidget(
        block, items, style, startCorner, highlightStyle, highlightSymbol, r);
  }

  public Bounds getItemsBounds(Optional<Integer> selected0, int offset0, int maxHeight) {
    int offset = Math.min(offset0, Saturating.saturatingSubUnsigned(items.length, 1));
    int start = offset;
    int end = offset;
    int height = 0;
    int it = offset;
    boolean cont = true;
    while (cont && it < items.length) {
      Item item = items[it];
      if (height + item.height() > maxHeight) {
        cont = false;
      } else {
        height += item.height();
        end += 1;
        it += 1;
      }
    }

    int selected = Math.min(selected0.orElse(0), items.length - 1);
    while (selected >= end) {
      height = Saturating.saturatingAdd(height, items[end].height());
      end += 1;
      while (height > maxHeight) {
        height = Saturating.saturatingSubUnsigned(height, items[start].height());
        start += 1;
      }
    }
    while (selected < start) {
      start -= 1;
      height = Saturating.saturatingAdd(height, items[start].height());
      while (height > maxHeight) {
        end -= 1;
        height = Saturating.saturatingSubUnsigned(height, items[end].height());
      }
    }
    return new Bounds(start, end);
  }

  @Override
  public void render(Rect area, Buffer buf, State state) {
    buf.setStyle(area, style);
    Rect listArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      listArea = innerArea;
    } else {
      listArea = area;
    }

    if (listArea.width() < 1 || listArea.height() < 1) {
      return;
    }

    if (items.length == 0) {
      return;
    }
    int listHeight = listArea.height();

    Bounds bounds = getItemsBounds(state.selected, state.offset, listHeight);
    int start = bounds.start();
    int end = bounds.end();
    state.offset = start;

    String highlightSymbol1 = highlightSymbol.orElse("");
    String blankSymbol = " ".repeat(new Grapheme(highlightSymbol1).width());

    int[] currentHeight = {0};
    boolean hasSelection = state.selected.isPresent();
    Ranges.range(
        state.offset,
        state.offset + end - start,
        i -> {
          Item item = items[i];
          int x;
          int y;
          if (startCorner == Corner.BottomLeft) {
            currentHeight[0] += item.height();
            x = listArea.left();
            y = listArea.bottom() - currentHeight[0];
          } else {
            x = listArea.left();
            y = listArea.top() + currentHeight[0];
            currentHeight[0] += item.height();
          }
          Rect itemArea = new Rect(x, y, listArea.width(), item.height());

          Style itemStyle = style.patch(item.style());
          buf.setStyle(itemArea, itemStyle);

          boolean isSelected = state.selected.isPresent() && state.selected.get() == i;
          tui.Spans[] lines = item.content().lines();
          for (int j = 0; j < lines.length; j++) {
            tui.Spans line = lines[j];
            String symbol;
            if (isSelected && (j == 0 || repeatHighlightSymbol)) {
              symbol = highlightSymbol1;
            } else {
              symbol = blankSymbol;
            }
            int elemX;
            int maxElementWidth;
            if (hasSelection) {
              Position p = buf.setStringn(x, y + j, symbol, listArea.width(), itemStyle);
              elemX = p.x();
              maxElementWidth = listArea.width() - (elemX - x);
            } else {
              elemX = x;
              maxElementWidth = listArea.width();
            }
            buf.setSpans(elemX, y + j, line, maxElementWidth);
          }
          if (isSelected) {
            buf.setStyle(itemArea, highlightStyle);
          }
        });
  }

  @Override
  public void render(Rect area, Buffer buf) {
    State state = new State(0, Optional.empty());
    render(area, buf, state);
  }

  public record Bounds(int start, int end) {}

  public static final class State {
    public int offset;
    public Optional<Integer> selected;

    public State(int offset, Optional<Integer> selected) {
      this.offset = offset;
      this.selected = selected;
    }

    public static State empty() {
      return new State(0, Optional.empty());
    }

    public void select(Optional<Integer> index) {
      selected = index;
      if (index.isEmpty()) {
        offset = 0;
      }
    }
  }

  public record Item(Text content, Style style) {
    public int height() {
      return content.height();
    }

    public int width() {
      return content.width();
    }
  }
}
